doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,752 @@
+
+/*
+ * Returns 1 if the SSL object is the default stream for the connection or 0
+ * otherwise. For (D)TLS all SSL objects are default. For QUIC only the first
+ * stream created for a connection is the default.
+ */
+int SSL_is_default_stream(SSL *s);
+
+/*
+ * If the event queue is not empty then this returns immediately and filss in
+ * ev with details of the next event. If the event queue is empty then it will
+ * attempt to read a packet from the BIO associated with the given SSL object,
+ * or if there is no associated BIO, or if the SSL object is NULL, then it will
+ * attempt to read a packet from the rbios associated with the SSL_EVENT_CTX and
+ * will then populate the event queue accordingly and fill ev with the first
+ * event. For DTLS this could fulfil the role of DTLSv1_listen() and create a
+ * "new connection" event only when the cookie has been verified.
+ * If after the above the event queue is still empty and SSL_EVENT_TYPE_TICK is
+ * not masked out then it will populate the event queue with any expired timer
+ * events for s if it is non NULL, or any SSL objects otherwise.
+ * If shouldblock is 1 and no events are available this function call will block
+ * until an event is available, or the timeout value has expired. If shouldblock
+ * is 1 and timeout is NULL then the function will block indefinitely until an
+ * event is available. If shouldblock is 0 then then function will always return
+ * immediately.
+ */
+int SSL_EVENT_CTX_get_next_event(SSL_EVENT_CTX *ectx, SSL *s,
+                                 int shouldblock, struct tm *timeout,
+                                 SSL_EVENT *ev);
+
+/*
+ * Add a pending new connection to the new connection queue in the SSL_EVENT_CTX.
+ * This will result in an SSL_EVENT_TYPE_NEW_CONNECTION event being issued. A
+ * subsequent call to SSL_accept will accept the connection and configure the
+ * SSL object for it. Applications should only need to do this for TLS because we
+ * don't otherwise know about the arrival of a new connection. Even for TLS this
+ * is optional (new connections can just be handled as they are now without the
+ * use of SSL_EVENT_CTX if desired). For QUIC and DTLS these new connections
+ * can be detected automatically when SSL_EVENT_CTX_get_next_event() processes a
+ * packet.
+ */
+int SSL_EVENT_CTX_add_new_connection(SSL_EVENT_CTX *ectx, BIO *rbio, BIO *wbio);
+
+/*
+ * Construct and free a new SSL_EVENT object
+ */
+SSL_EVENT *SSL_EVENT_new(void);
+void SSL_EVENT_free(SSL_EVENT *ev);
+
+
+/* Gives an integer value representing the type of event that occurred */
+uint64_t SSL_EVENT_get_event_type(SSL_EVENT *event);
+
+/*
+ * Gets the stream associated with a given event. In the case of "new connection"
+ * this will be NULL. In case the of "new stream" or "connection close" it will
+ * be the "default" stream (i.e. first stream) for the connection
+ */
+SSL *SSL_EVENT_get_stream(SSL_EVENT *event);
```

Comment:
@hlandau wrote:
> We might need other information on an event. For example, QUIC can associate an application error code with stream and connection close. For connection close, the application can even specify a UTF-8 reason string.
>
> This can be extended later by adding more functions on SSL_EVENT though without creating any technical debt, so this doesn't need to be dealt with immediately.

Absolutely. This is not intended to be a complete and fully formed API at this stage. I would fully expect us to expand it as we work further with it. Associating the error code and reason strings with the `SSL_EVENT` would make a lot of sense.

Reply:
I've expanded out the `SSL_EVENT` API to add functions for obtaining error codes and reason strings.